File: 백준/Gold/1806. 부분합/부분합.py
n, s = map(int, input().split())
nums = list(map(int, input().split()))
left, right = 0, 0
part_sum = 0
min_length = 1e9

while True:
  if part_sum >= s:
    min_length = min(min_length, right - left)
    part_sum -= nums[left]
    left += 1
  elif right == n:
    break
  else:
    part_sum += nums[right]
    right += 1
    
if min_length == 1e9:
  print(0)
else:
  print(min_length)

File: 백준/Gold/1806. 부분합/부분합.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int S = Integer.parseInt(st.nextToken());

        int[] nums = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        // S 입력 범위 100,000,000
        // 시간 제한 Java 1초
        // n^2 불가능, N 이하로 연산

        // DP(누적합), 투포인터 둘 다 가능할 것 같음
        int len = Integer.MAX_VALUE;
        int left = 0;
        int right = 0;
        int sum = 0;
        while (true) {
            if (sum >= S) {
                len = Math.min(len, right - left);
                sum -= nums[left++];
            } else {
                if (right == N) {
                    break;
                }
                sum += nums[right++];
            }
        }

        bw.write(len == Integer.MAX_VALUE ? "0\n" : (len + "\n"));
        bw.flush();
    }
}   
